snap/epoch.go

```diff
@@ -0,0 +1,288 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snap
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sort"
+	"strconv"
+)
+
+var (
+	ErrBadEpochExpression  = errors.New("invalid epoch expression")
+	ErrEpochOhSplat        = errors.New("0* is an invalid epoch")
+	ErrBadEpochNumber      = errors.New("epoch numbers must match [1-9][0-9]*")
+	ErrHugeEpochNumber     = errors.New("epoch numbers must be less than 2³²")
+	ErrBadEpochList        = errors.New("epoch read/write attributes must be lists of epoch numbers")
+	ErrEmptyEpochList      = errors.New("epoch list cannot be explicitly empty")
+	ErrEpochListNotSorted  = errors.New("epoch list must be in ascending order")
+	ErrNoEpochIntersection = errors.New("epoch read and write lists must have a non-empty intersection")
+)
+
+// An Epoch represents the ability of the snap to read and write its data. Most
+// developers need not worry about it, and snaps default to the 0th epoch, and
+// users are only offered refreshes to epoch 0 snaps. Once an epoch bump is in
+// order, there's a simplified expression they can use which should cover the
+// majority of the cases:
+//
+//   epoch: N
+//
+// means a snap can read/write exactly the Nth epoch's data, and
+//
+//   epoch: N*
+//
+// means a snap can additionally read (N-1)th epoch's data, which means it's a
+// snap that can migrate epochs (so a user on epoch 0 can get offered a refresh
+// to a snap on epoch 1*).
+//
+// If the above is not enough, a developer can explicitly describe what epochs a
+// snap can read and write:
+//
+//   epoch:
+//     read: [1, 2, 3]
+//     write: [1, 3]
+//
+// the read attribute defaults to the value of the write attribute, and the
+// write attribute defaults to the last item in the read attribute. If both are
+// unset, it's the same as not specifying an epoch at all (i.e. epoch: 0). The
+// lists must be in ascending order, and there must be a non-empty intersection
+// between them.
+//
+// Epoch numbers must be written as base 10 numbers, with no zero padding.
+type Epoch struct {
+	Read  []uint32 `yaml:"read"`
+	Write []uint32 `yaml:"write"`
+}
+
+// E returns the epoch represented by the expression s. It's meant for use in
+// testing, as it panics at the first sign of trouble.
+func E(s string) Epoch {
+	var e Epoch
+	if err := e.fromEpochString(s); err != nil {
+		panic(fmt.Errorf("%q: %v", s, err))
+	}
+	return e
+}
+
+// EpochZero returns the epoch represented by the expression "0".
+func EpochZero() Epoch {
+	return Epoch{Read: []uint32{0}, Write: []uint32{0}}
+}
+
+// IsNull checks whether this epoch looks like it's not been initialized.
+func (e Epoch) IsNull() bool {
```

Comment:
Can we please call this one 'Defined()`, inverting its logic?  We have a clear definition of something being null which would be better not to mix with high-level ideas as we have here.

Reply:
or drop it entirely if we go with the nil-is-zero

Reply:
dropped